[13.0][MIG] partner_email_check: Migration to 13.0

qgroulard opened the pull request.

denislour committed:
Add new module partner-email-check (#496)

* Add new module partner-email-check
Validate email input

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
Migrate module partner_email_check to version 11.0

Commit:
Remove copyright in init + ref code.

OCA-git-bot committed:
[ADD] icon.png

oca-travis committed:
[UPD] Update partner_email_check.pot

ntsirintanis committed:
[FIX][11.0] Make debugger record a debug message instead of error when importing validate_email in partner_email_check

OCA-git-bot committed:
partner_email_check 11.0.1.0.1

janverb committed:
Add normalization and optional uniqueness to partner_email_check

- partner_email_check now uses email-validator instead of
  validate_email
- Email addresses are normalized
- There is a setting to enforce uniqueness of partner email addresses
- There is a setting to check whether emails are
  deliverable (i.e. whether the domain resolves)

janverb committed:
[FIX] Fix partner_email_check conflict with portal, bump version

portal's tests set an invalid email address, `c@c`, that's rejected
by the current version of partner_email_check. If partner_email_check
is installed before portal's tests run then they fail.

To solve this, skip email validation if tests are running, unless a
key in the context is set to make the checks run anyway. That lets
tests in other addons opt in to the checks as well.

The previous update to partner_email_check didn't bump the version
number, so do that now.

oca-travis committed:
[UPD] Update partner_email_check.pot

OCA-git-bot committed:
partner_email_check 11.0.2.0.0

Commit:
[12.0][MIG] partner_email_check: Migration to 12.0

qgroulard committed:
[MIG] partner_email_check: Migration to 13.0

kos94ok-3D commented:
Not allowed duplicate partner with filled email so maybe need set `copy=False` or if `_should_filter_duplicates` remove `email` from `copy_data`

simahawk commented on partner_email_check/models/res_config_settings.py:
```diff
@@ -0,0 +1,43 @@
+
+    partner_email_check_filter_duplicates = fields.Boolean(
+        string="Filter duplicate partner email addresses",
+        help="Don't allow multiple partners to have the same email address.",
```
```suggestion
        help="Don't allow multiple partners to have the same email address.",
        config_parameter="partner_email_check_filter_duplicates"
```

simahawk commented on partner_email_check/models/res_config_settings.py:
```diff
@@ -0,0 +1,43 @@
+
+    partner_email_check_check_deliverability = fields.Boolean(
+        string="Check deliverability of email addresses",
+        help="Don't allow email addresses with providers that don't exist",
```
```suggestion
        help="Don't allow email addresses with providers that don't exist",
        config_parameter="partner_email_check_check_deliverability"
```
and trash get/set_values overrides ;)

BTW IMO this should be configured on the company, then you can get rid of the params.

simahawk commented on partner_email_check/models/res_partner.py:
```diff
@@ -0,0 +1,99 @@
+        ):
+            return emails
+        return ",".join(
+            self._normalize_email(email.strip()) for email in emails.split(",")
```
```suggestion
            self._normalize_email(email.strip()) for email in emails.split(",") if email.strip()
```

simahawk commented on partner_email_check/models/res_partner.py:
```diff
@@ -0,0 +1,99 @@
+            .sudo()
+            .get_param("partner_email_check_filter_duplicates", "False")
+        )
+        return conf == "True"
```
use `tools.str2bool` :wink:

simahawk commented on partner_email_check/tests/test_partner_email_check.py:
```diff
@@ -0,0 +1,147 @@
+from odoo.tools.misc import mute_logger
+
+
+class TestPartnerEmailCheck(TransactionCase):
```
pls use `SavepointCase` and use `cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))` to disable tracking globally.

rousseldenis commented:
@qgroulard Could you attend comments?

qgroulard commented:
> @qgroulard Could you attend comments?

Yes I am trying to squeeze this in today's agenda but I'm afraid it will be for another day.
I'll work on this asap.

qgroulard committed:
[FIX] Remove email from copy data if duplicates not allowed

[REF] Partner email chack params on company

[IMP] email_check

[IMP] test partner_email_check use SavepointCase

NL66278 commented on partner_email_check/models/res_partner.py:
```diff
@@ -24,14 +24,24 @@
+        if self._should_filter_duplicates():
+            for copy_vals in res:
+                if "email" in copy_vals:
+                    copy_vals["email"] = ""
```
Why not delete the value from the dictionary, or set it to False, so email will have a NULL value in the database.

NL66278 reviewed (approved):
:+1: LGTM

rousseldenis commented:
@simahawk Could you update your review?

simahawk commented on partner_email_check/models/res_partner.py:
```diff
@@ -0,0 +1,99 @@
+    def copy_data(self, default=None):
+        res = super(ResPartner, self).copy_data(default=default)
+        if self._should_filter_duplicates():
+            for copy_vals in res:
+                if "email" in copy_vals:
+                    del copy_vals["email"]
```
not a blocker but you could do:
```suggestion
            for copy_vals in res:
                copy_vals.pop("email", None)
```

simahawk commented on partner_email_check/models/res_partner.py:
```diff
@@ -0,0 +1,99 @@
+
+    @api.model
+    def email_check(self, emails):
+        if config["test_enable"] and not self.env.context.get(
```
do we really need this? What's the main reason?

simahawk commented on partner_email_check/tests/test_partner_email_check.py:
```diff
@@ -0,0 +1,141 @@
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
+        # Checks are disabled during tests unless this key is set
```
I don't get why we need this ctx key :thinking: 
Checks should be disabled by the flags set on the company.
If you want to keep them ON by default you can set them OFF in tests for the rest of the world by adding demo data :wink:

qgroulard replied:
This is actually the same comment as https://github.com/OCA/partner-contact/pull/989#discussion_r665111764 since this key is only used there.

I think I get it now...

There are actually 3 checks performed:
-  Check email unique: only if set on company;
-  Check deliverability: only if set on company, performed with https://pypi.org/project/email-validator/ thus can't run without checking emails syntax as well;
- Check syntax: currently always performed unless in test mode without this key (test_partner_email_check), performed with https://pypi.org/project/email-validator/

So I guess the idea is not to break other modules' tests since not having this context key set (and being in test mode) is the only way not to check emails syntax.

Maybe the solution would be to add a third flag for the syntax check.
It should be off by default otherwise other modules' test may fail. Thus users must be aware that behavior changes with v13 since beforehand this check used to always be performed.

What do you think ?

simahawk replied:
nope, the solution is simpler and I reported it above :wink: 

Just do this:

1. add demo data file
2. in that file call write function on all companies
3. set the flags to false
4. in your test class enable the flags

This way for the rest of the test suites (referred as "rest of the world" above) won't be impacted.

simahawk replied:
of course, the alternative is that you make the flag false by default and you state in the README that it should be enabled

NL66278 commented on partner_email_check/migrations/13.0.1.0.0/post-migration.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2021 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
```
Make it https:

qgroulard committed:
[IMP] Add partner email check syntax flag

simahawk reviewed (approved):
thanks for your changes!

simahawk commented:
/ocabot merge minor
